File: contracts/python/mirage_contracts/envelope.py
"""High-level validate/construct API for event and command envelopes.

This is the ONE place application code should call to accept an inbound
event or command: it validates envelope structure, checks the payload cap,
resolves and validates the payload against its registered schema, and
enforces the major-version compatibility rule (Step 1 / §Appendix C).
"""
from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass
from typing import Any

from mirage_contracts.errors import (
    EnvelopeValidationError,
    IntegrityMismatchError,
    PayloadTooLargeError,
    PayloadValidationError,
    UnknownCommandTypeError,
    UnknownEventTypeError,
    UnsupportedSchemaVersionError,
)
from mirage_contracts.registry import commands_registry, events_registry, parse_schema_version
from mirage_contracts.timestamps import now_rfc3339_ms
from mirage_contracts.ulid import generate_ulid

MAX_PAYLOAD_BYTES = 256 * 1024


def canonical_json_bytes(obj: Any) -> bytes:
    """Deterministic JSON encoding used for both size checks and integrity.sha256."""
    return json.dumps(obj, sort_keys=True, separators=(",", ":"), ensure_ascii=False).encode("utf-8")


def sha256_hex(obj: Any) -> str:
    return hashlib.sha256(canonical_json_bytes(obj)).hexdigest()


@dataclass(frozen=True)
class ValidatedEvent:
    envelope: dict
    payload: dict
    event_type: str
    major_version: int


@dataclass(frozen=True)
class ValidatedCommand:
    envelope: dict
    params: dict
    command_type: str
    major_version: int


def _envelope_errors(validator, instance: dict) -> list[str]:
    return [
        f"{'/'.join(str(p) for p in e.absolute_path) or '<root>'}: {e.message}"
        for e in sorted(validator.iter_errors(instance), key=lambda e: list(e.absolute_path))
    ]


def validate_event(instance: dict, *, supported_majors: set[int] | None = None) -> ValidatedEvent:
    """Validate a raw event dict end-to-end. Raises a typed ContractError subclass on failure.

    Rejection happens BEFORE any NATS publish or Elasticsearch write — callers
    must call this at the ingestion boundary, not after.
    """
    envelope_errors = _envelope_errors(events_registry.envelope_validator, instance)
    if envelope_errors:
        raise EnvelopeValidationError(envelope_errors)

    payload_bytes = len(canonical_json_bytes(instance["payload"]))
    if payload_bytes > MAX_PAYLOAD_BYTES:
        raise PayloadTooLargeError(payload_bytes)

    expected_sha256 = sha256_hex(instance["payload"])
    actual_sha256 = instance["integrity"]["sha256"]
    if expected_sha256 != actual_sha256:
        raise IntegrityMismatchError(expected=actual_sha256, actual=expected_sha256)

    event_type = instance["event_type"]
    major, _minor = parse_schema_version(instance["schema_version"])

    registered_majors = events_registry.supported_majors(event_type)
    if not registered_majors:
        raise UnknownEventTypeError(event_type)

    allowed = supported_majors if supported_majors is not None else registered_majors
    if major not in allowed:
        raise UnsupportedSchemaVersionError(event_type, instance["schema_version"], allowed)

    validator = events_registry.get_validator(event_type, major)
    assert validator is not None
    payload_errors = _envelope_errors(validator, instance["payload"])
    if payload_errors:
        raise PayloadValidationError(event_type, payload_errors)

    return ValidatedEvent(envelope=instance, payload=instance["payload"], event_type=event_type, major_version=major)


def validate_command(instance: dict, *, supported_majors: set[int] | None = None) -> ValidatedCommand:
    envelope_errors = _envelope_errors(commands_registry.envelope_validator, instance)
    if envelope_errors:
        raise EnvelopeValidationError(envelope_errors)

    command_type = instance["command_type"]
    major, _minor = parse_schema_version(instance["schema_version"])

    registered_majors = commands_registry.supported_majors(command_type)
    if not registered_majors:
        raise UnknownCommandTypeError(command_type)

    allowed = supported_majors if supported_majors is not None else registered_majors
    if major not in allowed:
        raise UnsupportedSchemaVersionError(command_type, instance["schema_version"], allowed)

    validator = commands_registry.get_validator(command_type, major)
    assert validator is not None
    params_errors = _envelope_errors(validator, instance["params"])
    if params_errors:
        raise PayloadValidationError(command_type, params_errors)

    return ValidatedCommand(envelope=instance, params=instance["params"], command_type=command_type, major_version=major)


def build_event(
    *,
    event_type: str,
    schema_version: str,
    payload: dict,
    source_id: str,
    sequence: int,
    actor_type: str,
    classification: str,
    case_id: str | None = None,
    session_id: str | None = None,
    event_id: str | None = None,
    event_time: str | None = None,
) -> dict:
    """Construct a well-formed event envelope dict (does not validate — call
    validate_event() on the result if you need to confirm it before publishing).
    """
    now = now_rfc3339_ms()
    return {
        "event_id": event_id or generate_ulid(),
        "event_type": event_type,
        "schema_version": schema_version,
        "event_time": event_time or now,
        "ingest_time": now,
        "case_id": case_id,
        "session_id": session_id,
        "source_id": source_id,
        "sequence": sequence,
        "actor_type": actor_type,
        "integrity": {"sha256": sha256_hex(payload)},
        "classification": classification,
        "payload": payload,
    }

File: contracts/typescript/src/envelope.ts
/**
 * High-level validate/construct API for event and command envelopes.
 * Mirrors contracts/python/mirage_contracts/envelope.py.
 */
import { createHash } from "node:crypto";
import type { ValidateFunction, ErrorObject } from "ajv";
import {
  EnvelopeValidationError,
  IntegrityMismatchError,
  PayloadTooLargeError,
  PayloadValidationError,
  UnknownCommandTypeError,
  UnknownEventTypeError,
  UnsupportedSchemaVersionError,
} from "./errors.js";
import { commandsRegistry, eventsRegistry, parseSchemaVersion } from "./registry.js";
import { generateUlid } from "./ulid.js";
import { nowRfc3339Ms } from "./timestamps.js";

export const MAX_PAYLOAD_BYTES = 256 * 1024;

export function canonicalJsonBytes(obj: unknown): Buffer {
  return Buffer.from(canonicalStringify(obj), "utf-8");
}

function canonicalStringify(obj: unknown): string {
  if (obj === null || typeof obj !== "object") return JSON.stringify(obj);
  if (Array.isArray(obj)) return `[${obj.map(canonicalStringify).join(",")}]`;
  const keys = Object.keys(obj as Record<string, unknown>).sort();
  const parts = keys.map((k) => `${JSON.stringify(k)}:${canonicalStringify((obj as Record<string, unknown>)[k])}`);
  return `{${parts.join(",")}}`;
}

export function sha256Hex(obj: unknown): string {
  return createHash("sha256").update(canonicalJsonBytes(obj)).digest("hex");
}

function formatErrors(errors: ErrorObject[] | null | undefined): string[] {
  return (errors ?? []).map((e) => `${e.instancePath || "<root>"}: ${e.message}`);
}

export interface ValidatedEvent {
  envelope: Record<string, unknown>;
  payload: Record<string, unknown>;
  eventType: string;
  majorVersion: number;
}

export interface ValidatedCommand {
  envelope: Record<string, unknown>;
  params: Record<string, unknown>;
  commandType: string;
  majorVersion: number;
}

export function validateEvent(instance: Record<string, unknown>, supportedMajors?: Set<number>): ValidatedEvent {
  const envelopeValidator: ValidateFunction = eventsRegistry.getEnvelopeValidator();
  if (!envelopeValidator(instance)) {
    throw new EnvelopeValidationError(formatErrors(envelopeValidator.errors));
  }

  const payload = instance.payload as Record<string, unknown>;
  const payloadBytes = canonicalJsonBytes(payload).length;
  if (payloadBytes > MAX_PAYLOAD_BYTES) {
    throw new PayloadTooLargeError(payloadBytes);
  }

  const expectedSha256 = sha256Hex(payload);
  const actualSha256 = (instance.integrity as Record<string, unknown>).sha256 as string;
  if (expectedSha256 !== actualSha256) {
    throw new IntegrityMismatchError(actualSha256, expectedSha256);
  }

  const eventType = instance.event_type as string;
  const [major] = parseSchemaVersion(instance.schema_version as string);

  const registeredMajors = eventsRegistry.supportedMajors(eventType);
  if (registeredMajors.size === 0) {
    throw new UnknownEventTypeError(eventType);
  }
  const allowed = supportedMajors ?? registeredMajors;
  if (!allowed.has(major)) {
    throw new UnsupportedSchemaVersionError(eventType, instance.schema_version as string, [...allowed]);
  }

  const payloadValidator = eventsRegistry.getValidator(eventType, major)!;
  if (!payloadValidator(payload)) {
    throw new PayloadValidationError(eventType, formatErrors(payloadValidator.errors));
  }

  return { envelope: instance, payload, eventType, majorVersion: major };
}

export function validateCommand(instance: Record<string, unknown>, supportedMajors?: Set<number>): ValidatedCommand {
  const envelopeValidator = commandsRegistry.getEnvelopeValidator();
  if (!envelopeValidator(instance)) {
    throw new EnvelopeValidationError(formatErrors(envelopeValidator.errors));
  }

  const commandType = instance.command_type as string;
  const [major] = parseSchemaVersion(instance.schema_version as string);

  const registeredMajors = commandsRegistry.supportedMajors(commandType);
  if (registeredMajors.size === 0) {
    throw new UnknownCommandTypeError(commandType);
  }
  const allowed = supportedMajors ?? registeredMajors;
  if (!allowed.has(major)) {
    throw new UnsupportedSchemaVersionError(commandType, instance.schema_version as string, [...allowed]);
  }

  const params = instance.params as Record<string, unknown>;
  const paramsValidator = commandsRegistry.getValidator(commandType, major)!;
  if (!paramsValidator(params)) {
    throw new PayloadValidationError(commandType, formatErrors(paramsValidator.errors));
  }

  return { envelope: instance, params, commandType, majorVersion: major };
}

export interface BuildEventArgs {
  eventType: string;
  schemaVersion: string;
  payload: Record<string, unknown>;
  sourceId: string;
  sequence: number;
  actorType: string;
  classification: string;
  caseId?: string | null;
  sessionId?: string | null;
  eventId?: string;
  eventTime?: string;
}

export function buildEvent(args: BuildEventArgs): Record<string, unknown> {
  const now = nowRfc3339Ms();
  return {
    event_id: args.eventId ?? generateUlid(),
    event_type: args.eventType,
    schema_version: args.schemaVersion,
    event_time: args.eventTime ?? now,
    ingest_time: now,
    case_id: args.caseId ?? null,
    session_id: args.sessionId ?? null,
    source_id: args.sourceId,
    sequence: args.sequence,
    actor_type: args.actorType,
    integrity: { sha256: sha256Hex(args.payload) },
    classification: args.classification,
    payload: args.payload,
  };
}
